GPU/Common/DepthRaster.cpp

```diff
@@ -0,0 +1,479 @@
+	verts[0].x = vertsSub[0].x;
+	verts[0].y = vertsSub[0].y;
+	verts[0].z = vertsSub[0].z;
+	verts[1].x = vertsSub[2].x;
+	verts[1].y = vertsSub[2].y;
+	verts[1].z = vertsSub[2].z;
+	verts[2].x = vertsSub[1].x;
+	verts[2].y = vertsSub[1].y;
+	verts[2].z = vertsSub[1].z;
+
+	// use fixed-point only for X and Y.  Avoid work for Z and W.
+	int startX = std::max(std::min(std::min(verts[0].x, verts[1].x), verts[2].x), tileStartX);
+	int endX = std::min(std::max(std::max(verts[0].x, verts[1].x), verts[2].x) + 1, tileEndX);
+
+	int startY = std::max(std::min(std::min(verts[0].y, verts[1].y), verts[2].y), tileStartY);
+	int endY = std::min(std::max(std::max(verts[0].y, verts[1].y), verts[2].y) + 1, tileEndY);
+	if (endX == startX || endY == startY) {
+		// No pixels
+		return;
+	}
+	// TODO: Cull really small triangles here.
+
+	// Fab(x, y) =     Ax       +       By     +      C              = 0
+	// Fab(x, y) = (ya - yb)x   +   (xb - xa)y + (xa * yb - xb * ya) = 0
+	// Compute A = (ya - yb) for the 3 line segments that make up each triangle
+	int A0 = verts[1].y - verts[2].y;
+	int A1 = verts[2].y - verts[0].y;
+	int A2 = verts[0].y - verts[1].y;
+
+	// Compute B = (xb - xa) for the 3 line segments that make up each triangle
+	int B0 = verts[2].x - verts[1].x;
+	int B1 = verts[0].x - verts[2].x;
+	int B2 = verts[1].x - verts[0].x;
+
+	// Compute C = (xa * yb - xb * ya) for the 3 line segments that make up each triangle
+	int C0 = verts[1].x * verts[2].y - verts[2].x * verts[1].y;
+	int C1 = verts[2].x * verts[0].y - verts[0].x * verts[2].y;
+	int C2 = verts[0].x * verts[1].y - verts[1].x * verts[0].y;
+
+	// Compute triangle area
+	int triArea = A0 * verts[0].x + B0 * verts[0].y + C0;
+	if (triArea <= 0) {
+		// Too small to rasterize or backface culled
+		// NOTE: Just disabling this check won't enable two-sided rendering.
+		// Since it's not that common, let's just queue the triangles with both windings.
+		return;
+	}
+
+	int rowIdx = (startY * stride + startX);
+	int col = startX;
+	int row = startY;
+
+	// Calculate slopes at starting corner.
+	int alpha0 = (A0 * col) + (B0 * row) + C0;
+	int beta0 = (A1 * col) + (B1 * row) + C1;
+	int gamma0 = (A2 * col) + (B2 * row) + C2;
+
+	float oneOverTriArea = (1.0f / float(triArea));
+
+	float zz[3];
+	zz[0] = (float)verts[0].z;
+	zz[1] = (float)(verts[1].z - verts[0].z) * oneOverTriArea;
+	zz[2] = (float)(verts[2].z - verts[0].z) * oneOverTriArea;
+
+	// END triangle setup.
+
+	// Incrementally compute Fab(x, y) for all the pixels inside the bounding box formed by (startX, endX) and (startY, endY)
+	for (int r = startY; r < endY; r++,
+		row++,
+		rowIdx += stride,
+		alpha0 += B0,
+		beta0 += B1,
+		gamma0 += B2)
+	{
+		int idx = rowIdx;
+
+		// Restore row steppers.
+		int alpha = alpha0;
+		int beta = beta0;
+		int gamma = gamma0;
+
+		for (int c = startX; c < endX; c++,
+			idx++,
+			alpha += A0,
+			beta += A1,
+			gamma += A2)
+		{
+			int mask = alpha >= 0 && beta >= 0 && gamma >= 0;
```

Comment:
A classic trick is `int mask = (alpha | beta | gamma);`, followed by `if(mask < 0) continue;`.
Extends to SIMD by `_mm_movemask_ps` to check the 4 signs (on x86, not sure about NEON).

Reply:
Addendum: this is what SoftGPU does for NEON:
https://github.com/hrydgard/ppsspp/blob/53cb014c076ccf4d56b26e17bfe7b79802c27a79/GPU/Software/Rasterizer.cpp#L946